include/stl2/view/filter.hpp

```diff
@@ -0,0 +1,253 @@
+// cmcstl2 - A concept-enabled C++ standard library
+//
+//  Copyright Eric Niebler 2017
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/caseycarter/cmcstl2
+//
+#ifndef STL2_VIEW_FILTER_HPP
+#define STL2_VIEW_FILTER_HPP
+
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/iterator/concepts.hpp>
+#include <stl2/detail/range/access.hpp>
+#include <stl2/detail/range/concepts.hpp>
+#include <stl2/view/all.hpp>
+#include <stl2/view/view_interface.hpp>
+#include <stl2/optional.hpp>
+#include <stl2/detail/non_propagating_cache.hpp>
+#include <stl2/detail/functional/invoke.hpp>
+#include <stl2/detail/algorithm/find_if.hpp>
+#include <stl2/detail/semiregular_box.hpp>
+#include <stl2/detail/view/view_closure.hpp>
+
+STL2_OPEN_NAMESPACE {
+	namespace detail {
+		template <InputRange R>
+		struct begin_iterator_cache_ {
+			mutable detail::non_propagating_cache<iterator_t<R>> value_;
+			constexpr decltype(auto) get() const noexcept
+			{ return (value_); }
+		};
+		template <InputRange R>
+		requires InputRange<const R> && !Same<iterator_t<R>, iterator_t<const R>>
+		struct begin_iterator_cache_<R> {
+			mutable detail::non_propagating_cache<iterator_t<R>> value_;
+			mutable detail::non_propagating_cache<iterator_t<const R>> cvalue_;
+			constexpr decltype(auto) get() noexcept
+			{ return (value_); }
+			constexpr decltype(auto) get() const noexcept
+			{ return (cvalue_); }
+		};
+	} // namespace detail
+
+	namespace ext {
+		template </*InputRange*/ class R, IndirectUnaryPredicate<iterator_t<R>> Pred>
+		requires View<R>
+		class filter_view : view_interface<filter_view<R, Pred>> {
+		private:
+			R base_;
+			detail::semiregular_box<Pred> pred_;
+			detail::begin_iterator_cache_<R> cache_;
+			template <bool Const>
+			class __iterator;
+			template <bool Const>
+			class __sentinel;
+		public:
+			filter_view() = default;
+
+			constexpr filter_view(R base, Pred pred)
+			: base_(std::move(base)), pred_(std::move(pred)) {}
+
+			template <InputRange O>
+			requires (std::is_lvalue_reference_v<O> || View<__f<O>>) &&
+				Constructible<R, all_view<O>>
+			constexpr filter_view(O&& o, Pred pred)
+			: base_(view::all(std::forward<O>(o))), pred_(std::move(pred)) {}
+
+			constexpr R base() const
+			{ return base_; }
+
+			using iterator = __iterator<false>;
+			using sentinel = __sentinel<false>;
+			using const_iterator = __iterator<true>;
+			using const_sentinel = __sentinel<true>;
+
+			constexpr iterator begin()
+			{
+				if(!cache_.get())
+					cache_.get() = find_if(base_, std::ref(pred_.get()));
+				return iterator{*this, *cache_.get()};
+			}
+
+			// Template to work around https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82507
+			template <class ConstR = const R>
+			constexpr const_iterator begin() const
```

Comment:
I do too. I just haven't changed it yet.

Reply:
@CaseyCarter I can't think of any legitimate situation to use such an adaptor.  Does it have an use case?

FWIW I would be fine with it being added to the tests so that we can make sure that such an adaptor can be written and play a bit with it but adding it to the standard library makes me slightly uncomfortable.

Reply:
> I can't think of any legitimate situation to use such an adaptor. Does it have an use case?

You want to pass a view to an old library function that believes that all ranges are container-like, in that they can be passed by `const&` and iterated over. I think this is the use case that causes people to stumble over non-`const`-iterable views in the first place.

Reply:
Would a `view::input` satisfy that use case? I mean, what are the chances of the code expecting a `const&`-container is expecting something more than an `InputView` ? 

For example, lots of code expecting `const&` containers do not expect them to be single pass, but many views are single pass. 

Just because users can make a view `const`-iterable does not mean that the resulting view is useful. I've been biten by this a lot, the answer always was the simplest one: either fix the old code by removing `const` in the argument, or reimplementing it.

Reply:
> Would a view::input satisfy that use case? I mean, code expecting a const&-container typically expects something more than an InputView.

True; the container requirements specify forward iterators.

Reply:
I bet people are going to be angry at this, but broken code not compiling is a feature.